motionplan/tpSpaceRRT.go

```diff
@@ -788,51 +792,65 @@ func (mp *tpSpaceRRTMotionPlanner) setupTPSpaceOptions() {
 		goalMetricConstructor: defaultGoalMetricConstructor,
 	}
 
-	for _, curPtg := range mp.tpFrame.PTGSolvers() {
-		tpOpt.distOptions[curPtg] = mp.make2DTPSpaceDistanceOptions(curPtg)
+	mp.algOpts = tpOpt
+}
+
+func (mp *tpSpaceRRTMotionPlanner) ptgSolutionAndMetric(ptg tpspace.PTGSolver,
+	nearestPose, randPosNodePose spatialmath.Pose,
+) (*ik.Solution, ik.StateMetric, error) {
+	relPose := spatialmath.PoseBetween(nearestPose, randPosNodePose)
+	targetFunc := mp.algOpts.goalMetricConstructor(relPose)
+	seedDist := relPose.Point().Norm()
+	seed := tpspace.PTGIKSeed(ptg)
+	dof := ptg.DoF()
+	if seedDist < dof[1].Max {
+		seed[1].Value = seedDist
 	}
 
-	mp.algOpts = tpOpt
+	solutionChan := make(chan *ik.Solution, 1)
+	err := ptg.Solve(context.Background(), solutionChan, seed, targetFunc, 0)
+
+	var solution *ik.Solution
+	select {
+	case solution = <-solutionChan:
+	default:
+	}
+
+	return solution, targetFunc, err
 }
 
 // make2DTPSpaceDistanceOptions will create a plannerOptions object with a custom DistanceFunc constructed such that
 // distances can be computed in TP space using the given PTG.
-func (mp *tpSpaceRRTMotionPlanner) make2DTPSpaceDistanceOptions(ptg tpspace.PTGSolver) *plannerOptions {
+// Also returns a pointer to a sync.Map of nearest poses -> ik.Solution so the (expensive to compute) solution can be reused.
+func (mp *tpSpaceRRTMotionPlanner) make2DTPSpaceDistanceOptions(ptg tpspace.PTGSolver) (*plannerOptions, *sync.Map) {
+	m := sync.Map{}
 	opts := newBasicPlannerOptions(mp.frame)
 	segMetric := func(seg *ik.Segment) float64 {
+		distance := math.Inf(1)
 		// When running NearestNeighbor:
 		// StartPosition is the seed/query
 		// EndPosition is the pose already in the RRT tree
 		if seg.StartPosition == nil || seg.EndPosition == nil {
-			return math.Inf(1)
-		}
-		var targetFunc ik.StateMetric
-		relPose := spatialmath.PoseBetween(seg.EndPosition, seg.StartPosition)
-		seedDist := relPose.Point().Norm()
-		targetFunc = mp.algOpts.goalMetricConstructor(relPose)
-		seed := tpspace.PTGIKSeed(ptg)
-		dof := ptg.DoF()
-		if seedDist < dof[1].Max {
-			seed[1].Value = seedDist
-		}
-		solutionChan := make(chan *ik.Solution, 1)
-		err := ptg.Solve(context.Background(), solutionChan, seed, targetFunc, 0)
-		var closeNode *ik.Solution
-		select {
-		case closeNode = <-solutionChan:
-		default:
+			return distance
 		}
-		if err != nil || closeNode == nil {
-			return math.Inf(1)
+		solution, targetFunc, err := mp.ptgSolutionAndMetric(ptg, seg.EndPosition, seg.StartPosition)
+
+		if err != nil || solution == nil {
+			return distance
 		}
-		pose, err := ptg.Transform(closeNode.Configuration)
+
+		pose, err := ptg.Transform(solution.Configuration)
 		if err != nil {
-			return math.Inf(1)
+			return distance
 		}
-		return targetFunc(&ik.State{Position: pose})
+
+		distance = targetFunc(&ik.State{Position: pose})
```

Comment:
Looking at this again, I think we could probably just use `solution.Score` rather than recalculating. Then we wouldn't need to return `targetFunc` at all. Not sure why we didn't already do that.

Reply:
It's not obvious to me why that is the case but I've refactored the code to no longer use `targetFunc` & to instead use `solution.Score` in all cases I changed. Let me know if I have misunderstood the meaning of this comment.

Reply:
While this change appears to pass tests, it also appears to make the motion test performance much worse.

Reply:
Does the current state of the PR implement this suggestion correctly?

Reply:
Yes, I believe so